Foundation/CPCache.j

```diff
@@ -0,0 +1,350 @@
+
+    A CPCache object is a collection-like container, or cache, that stores key-value pairs,
+    similar to the CPDictionary class. Developers often incorporate caches to temporarily
+    store objects with transient data that are expensive to create.
+
+    Reusing these objects can provide performance benefits, because their values do not have to be recalculated.
+    However, the objects are not critical to the application and can be discarded if memory is tight.
+    If discarded, their values will have to be recomputed again when needed.
+ */
+@implementation CPCache : CPObject
+{
+    CPDictionary            _items              @accessors(readonly);  // TOFIX: delete accessors which is only needed for test
+    int                     _currentPosition;
+    int                     _countLimit;
+    int                     _totalCostLimit;
+    id <CPCacheDelegate>    _delegate;
+
+    // Set of bit to determine which delegate has to respond
+    unsigned                _implementedDelegateMethods;
+}
+
+
+#pragma mark Initialization
+
+/*!
+    Initializes the cache with default values
+    @return the initialized cache
+*/
+- (id)init
+{
+    self = [super init];
+
+    if (self)
+    {
+        _items = [[CPDictionary alloc] init];
+        _currentPosition = 0;
+        _countLimit = 0;
+        _totalCostLimit = 0;
+        _delegate = nil;
+    }
+
+    return self;
+}
+
+
+#pragma mark Managing cache
+
+/*!
+    Returns the object which correspond to the given key
+    @param aKey the key for the object's entry
+    @return the object for the entry
+*/
+- (id)objectForKey:(id)aKey
+{
+    return [[_items objectForKey:aKey] object];
+}
+
+/*!
+    Adds an object with default cost into the cache.
+    @param anObject the object to add in the cache
+    @param aKey the object's key
+*/
+- (void)setObject:(id)anObject forKey:(id)aKey
+{
+    [self setObject:anObject forKey:aKey cost:0];
+}
+
+/*!
+    Adds an object with a cost into the cache.
+    @param anObject the object to add in the cache
+    @param aKey the object's key
+    @param aCost the object's cost
+*/
+- (void)setObject:(id)anObject forKey:(id)aKey cost:(int)aCost
+{
+    // Check if the key already exist
+    if ([_items objectForKey:aKey])
+        [self removeObjectForKey:aKey];
+
+    // Add object
+    [_items setObject:[_CPCacheItem cacheItemWithObject:anObject cost:aCost position:++_currentPosition] forKey:aKey];
+
+    // Clean cache to satisfy condition (< totalCostLimit & < countLimit) if necessary
+    [self _cleanCache];
+}
+
+/*!
+    Removes the object from the cache for the given key.
+    @param aKey the key of the object to be removed
+*/
+- (void)removeObjectForKey:(id)aKey
+{
+    // Call delegate method to warn that the object is going to be removed
+    if (_implementedDelegateMethods & CPCacheDelegate_cache_WillEvictObject)
```

Comment:
we normally create private methods to send call to delegates.

for instance:

```objj
- (void)_sendDelegateWillEvictObjectForKey:(CSPtring)aKey
{
    if (_implementedDelegateMethods & CPCacheDelegate_cache_WillEvictObject)
        [_delegate cache:self willEvictObject:[[_items objectForKey:aKey] object]];
}
```

then 

```objj
- (void)removeObjectForKey:(id)aKey
{
    [self _sendDelegateWillEvictObjectForKey:aKey];
    [_items removeObjectForKey:aKey];
}
```